Export `as_opt_<type>` helper functions

rnag opened the pull request:
Closes #4

rnag committed:
export `as_opt_<type>` helper functions

rnag committed:
use `try_from(v).ok()` instead

rnag committed:
add `examples/as_opt_bool`

rnag committed:
add/update examples

rnag committed:
add/update examples and docs

rnag committed:
update usage in docs

rnag committed:
docs: add section on "Optionals"

rnag committed:
cleanup doc comments

rnag committed:
change examples path

rnag committed:
reformat docs

corneliusroemer commented on examples/demo.rs:
```diff
@@ -26,7 +30,8 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         "numAttempts": "",
-        "grade": "81"
+        "grade": "81",
+        "notes": ""
```
I'm unsure here: should empty strings not be deserialized to `Some(String::new())`?

Is there something that cannot be unambiguously deserialized into a String? Only that should be `None`.

rnag replied:
A very good point! I wasn't sure of the expected behavior either, but that reasoning makes sense to me. I updated logic so that `as_opt_string` deserializes empty strings to `Some("")` rather than `None`.

corneliusroemer commented on examples/optionals/as_bool.rs:
```diff
@@ -0,0 +1,152 @@
+    }"#;
+
+    let m: Msg = serde_json::from_str(data).unwrap();
+    assert_eq!(m.timestamp, Some(false));
```
This should be `None`, shouldn't it? When parsing into an Option, only unambiguiously parseable strings should be turned into a Some variant. At least that would be my expectation. Or do we follow Javascript parsing logic, that is much laxer and casts eagerly?
```suggestion
    assert_eq!(m.timestamp, None);
```

Maybe we should write out all the edge cases and how they should treat it.

rnag replied:
Done @corneliusroemer ! Thanks for bringing this up. I added a check for all "falsy" string values explicitly. All other (invalid) string values should be deserialized to `None` as suggested.

corneliusroemer commented on examples/optionals/as_bool.rs:
```diff
@@ -0,0 +1,152 @@
+#[derive(Clone, Debug, Deserialize)]
+pub struct Msg {
+    #[serde(deserialize_with = "as_opt_bool")]
+    pub timestamp: Option<bool>,
```
I'm confused why you associate `timestamp` and `Option<bool>`. A timestamp is not boolean, so why not use something more typically boolean, like `truthy` or something that doesn't have any baggage like `foo`.

rnag replied:
done! agreed, my process was in cloning examples and this resulted in retaining the same named field `timestamp` for all examples, however I agree that it doesn't make sense in examples for `bool` fields. I renamed the field to `archived` so that we can be unambiguous.

corneliusroemer commented on examples/optionals/as_u64.rs:
```diff
@@ -0,0 +1,98 @@
+
+    let m: Msg = serde_json::from_str(data).unwrap();
+    trace!("  {m:?}");
+    assert_eq!(m.timestamp, Some(1));
```
I think this could arguably be `None`. To round or not to round...

rnag replied:
Agreed! When I searched online, I found that the rounding logic can be considerably more complex, as it relies on if a number to the left of the decimal point is even or odd. This makes rounding `0.5` ambiguous, as `0` is neither odd or even (or maybe both?).

For now, I wouldn't say that it's necessarily *wrong* to round 0.5 to 1. That is, it appears that the rounding logic is at least consistent, so that it is expected that 0.5 would round upwards to 1 (same with 3.5 or 4.5 for example). 

However, I'm definitely open to exploring this scenario more in detail, and seeing what differing thoughts and opinions might arise on this issue with rounding numbers in this case.

corneliusroemer commented on src/de_impl_opt.rs:
```diff
@@ -0,0 +1,442 @@
+///   * a `null` value.
+///   * any *de-serialization* errors.
+///     * ex. a string is non-empty and not a valid numeric value.
+///     * ex. the signed value `i64` represents a *negative* number.
```
```suggestion
///     * ex. the signed value `i64` represents a *negative* number.
///     * ex. float `f64` represents a *negative* number `< -0.5`, or `NaN`.
```

corneliusroemer commented on src/de_impl_opt.rs:
```diff
@@ -0,0 +1,442 @@
+/// > Note: the pattern matching is *case insensitive*, so `YES` or `yes`
+/// > works just the same.
+///
+/// These are the following "truthy" string values that result in a
```
For completeness, also the `falsy` values need to be defined - since we're moving to three valued logic now. This seems to not have been required previously as everything that was not `truthy` was `falsy`. But now the formerly `falsy` need to be split into `falsy` and `None`

rnag replied:
Done @corneliusroemer ! Thanks for bringing this up. I have added a doc section for the function `as_opt_bool` documenting the "falsy" string values, as suggested.

corneliusroemer reviewed (commented):
There seems to be some differing assumptions on what should be None and what shouldn't.

I find the `examples` section hard to read. Rather than having many examples, it may be better to have unit tests - and examples (doc tests) in `src/de_impl_opt.rs` doc strings.

Somewhere there should be a dense table/section where the behaviour can be seen at a glance - possibly in the doc string itself. I like the `# Returns` section, explaining behaviour, but this could be complemented with more comprehensive cases for each of the input types `str`, `i64`, `f64`, `u64`.

rnag committed:
`as_opt_string` should deserialize empty strings `to Some(String::new())` rather than `None`

rnag committed:
as_bool: rename `timestamp` to `archived` in examples to resolve ambiguity

rnag committed:
Update src/de_impl_opt.rs

Co-authored-by: Cornelius Roemer <cornelius.roemer@gmail.com>

rnag committed:
`as_opt_bool`: deserialize only "truthy" or "falsy" values to `Some(bool)`, and others to `None`

rnag commented:
> There seems to be some differing assumptions on what should be None and what shouldn't.
> 
> I find the `examples` section hard to read. Rather than having many examples, it may be better to have unit tests - and examples (doc tests) in `src/de_impl_opt.rs` doc strings.

Hi @corneliusroemer ! Thanks for bringing this up. I totally agree, that at some point it would be worth it to refactor it to move example cases into doc tests & unit tests more explicitly, that way it also runs automatically in the GitHub Actions workflow (CI) and also when deploying new versions.

I've created a new issue to track effort on this separately, as part of #7 .

For now, as and when when time allows, please take a look at recent changes and commits I have added as part of this PR, to confirm if it resolves or addresses all issues based on your notes. I will hold off on merging the changes until then. 

Thanks, and have a great Monday 🎉 !

rnag committed:
fix: resolve conflict in docs to agree with updated logic

josephwynn-sc commented:
I just stumbled upon this crate and the first thing I wanted was `as_opt_*` - is there a chance this PR would be merged in the near future?

rnag commented:
@josephwynn-sc  Yes, totally! Sorry for the long wait, just coming back after a hiatus.

I'll look into it this weekend, run some tests with `rust` -- now that I've got a new laptop and got Rust set up -- and if everything looks good, I'll go ahead and merge this PR. Thanks!

rnag committed:
add tests

rnag committed:
add moar tests

rnag committed:
add  tests

rnag committed:
add tests for de_impl_opt.rs

rnag committed:
add tests for de_impl_opt.rs

rnag committed:
Add release notes for `v0.5.0`

rnag committed:
fix `clippy` errors

rnag commented:
@corneliusroemer @josephwynn-sc Done -- PR has been merged, and exported `opt_` functions are available as of `v0.5.0` 🎉 

Feel free to check it out and let me know! 😃
